File: BOJ_11724.py
#https://www.acmicpc.net/problem/11724
# -*- coding: utf-8 -*-
# UTF-8 encoding when using korean
import sys
sys.setrecursionlimit(10000) #need for runtime error


def dfs(i):
	visited[i]=1
	for j in arr[i]:
		if visited[j] == 0 :
			#visited[j]=1
			dfs(j)

#N,M=map(int, input().split())
N,M=map(int, sys.stdin.readline().split())
arr=[[] for _ in range(N+1)]
visited=[0]*(N+1)
cnt=0

for i in range(M):
	#s,e=map(int, input().split())
	s,e=map(int, sys.stdin.readline().split()) #resolve time exceed error
	arr[s].append(e)
	arr[e].append(s)
	
for i in range(1,N+1):
	if visited[i] == 0 :
		cnt+=1
		#visited[i]=1
		dfs(i)

print(cnt)

File: BOJ_11724.cpp
//https://www.acmicpc.net/problem/11724
#include <iostream>
using namespace std;

int arr[1000+10][1000+10];
int visited[1000+10];
int N,M;
int cnt;

void dfs(int i){
	for(int j=1;j<=N;j++){
		if(arr[i][j]==1 && visited[j]!=1){
			visited[j]=1;
			dfs(j);
		}
	}
}

int main() {
	cin>>N>>M;
	for(int i=0;i<M;i++){
		int s,e;
		cin>>s>>e;
		arr[s][e]=1;
		arr[e][s]=1;
	}
	
	for(int i=1;i<=N;i++){
		if(visited[i]==0){
			cnt++;
			visited[i]=1;
			dfs(i);
		}
	}
	
	cout<<cnt;
	return 0;
}